Updates guidance on performance - font loading (#378)
springernature/frontend-playbook: performance/font-loading.md

# Font Loading

## Are you sure you need to use webfonts?

> It's worth asking this question initially. What benefit do you believe a webfont will give you and is it worth the cost in page performance? Afterall, system fonts are battle tested and were produced with legibility in mind.

Reviewer: I love the fact that this is the first item. I would suggest rephrasing it so the title is not a question, to something like:

> ## When to use webfonts
> 
> Before using webfonts it's worth considering what benefits webfonts provide, and what impact they may have on things like accessibility or page speed. For example:
> 
> * System fonts are battle tested and were produced with legibility in mind. Using webfonts may reduce the readability of the site, especially for people with dyslexia.
> * The negative impact of webfonts (additional requests and increased page weight) on page load times may cause a page to go from "very slow" to "unusable", especially in markets of strategic importance for Springer Nature like China. This will in turn impact usage numbers.

Author: Thanks, great suggestion. I have reworded it a bit more as well to try and account for the fact that a person or team might have their own specific reasons for why the webfont might benefit their project, and also to communicate that it is advised to weigh up the pros and cons against each to ensure on balance the pros win.

## Implementation

> We do not prescribe a particular implementation. Each team should  determine which implementation is best suited for them. However, before selecting an implementation a team should be mindful of our working practises such as:

Reviewer 2: ```suggestion
We do not prescribe a particular implementation. Each team should determine which implementation is best suited for them. However, before selecting an implementation a team should be mindful of our working practices such as:
```
spotted a typo and double space :)

Author: Thank you!

- [Graded Browser Support](https://github.com/springernature/frontend-playbook/blob/main/practices/graded-browser-support.md)
  - Is the implementation supported by all "enhanced" browsers? If not, is there a good justification for selecting it anyway?
  - Does the implementation work in conjunction with the Cut The Mustard technique?
- [Progressive Enhancement](https://github.com/springernature/frontend-playbook/blob/main/practices/progressive-enhancement.md)
  - This is not meant to force you to choose a non-js implementation, but rather for you to ensure you consider what will happen for users in varying circumstances from the beginning.
- [Our Performance Goals](https://github.com/springernature/frontend-playbook/blob/main/performance/performance-checklist.md)
  - What impact will the implementation have on the page performance of your site? 
  
## Quick Tips
Here are some quick suggestions for techniques that may prove useful:

### Understand the different user experiences in font loading: FOIT, FOUT and FOIT

> - FOIT means flash of invisible text. When web fonts are loading, we hide text so users don’t see anything. We only show the text when web fonts are loaded.

Reviewer: > When web fonts are loading, we hide text so users don’t see anything.

Maybe stick a "while" in there, in case people less familiar with technical details end up on this page, something like:

> Text on the page is hidden while webfonts are loading, so users will (or may??) see a page with images but no text. We only show the text when [...]

- FOUT means flash of unstyled text. When web fonts are loading, we show users a system font. When the web font gets loaded, we change the text back to the desired web font.

> - FOFT means flash of faux text. Only load the roman character set initially and load the others later for less page render blocking. "Faux" refers to the fact the browser fakes italics and bold before those character sets have loaded.

Reviewer: Suggestion: make definitions a bit clearer by putting "flash of [...] text" in italics, e.g.:

> FOIT means _flash of invisible text_.

We believe that FOIT should be avoided as it is the worst experience for the user.

### Understand the different implementations available

We highly recommend reading [Zach Leat's comprehensive guide](https://www.zachleat.com/web/comprehensive-webfonts/) to get a good understand of the different techniques available. There is also a good [follow-up article](https://css-tricks.com/the-best-font-loading-strategies-and-how-to-execute-them/) that summarises Zach's findings.

### Serve WOFF and WOFF2 font formats

> WOFF2 has the superior compression. However, unfortunately it does not cover all of our Enhanced browsers at the time of writing. Therefore consider serving both formats. 

Reviewer: Add a link to caniuse, for example:

> [...] unfortunately [it does not cover all of our Enhanced browsers](https://caniuse.com/woff2) at the time of writing [...]

# Font Loading

## When to use webfonts

> Before using webfonts it's worth understanding what benefits you believe a webfont will provide your product and assess those against the potential impact they may have on things like accessibility or page speed. For example:

Reviewer 3: ```suggestion
Before using webfonts it's worth defining what benefits you believe a webfont will provide your product, and assessing those against the potential impact they may have on things like accessibility or page speed. For example:
```

Author: Have reworked this section now so no need to address this suggestion. Have committed your other suggestions though. Thanks!

> - System fonts are extensively tested and produced with legibility in mind. Using webfonts may reduce the readability of the site, especially for people with dyslexia.

Reviewer 3: ```suggestion
- System fonts are extensively tested and produced with legibility in mind. Using webfonts may reduce the readability of the site, for example for people with dyslexia, or low vision.
```

We believe that FOIT should be avoided as it is the worst experience for the user.

### Understand the different implementations available

We highly recommend reading [Zach Leat's comprehensive guide](https://www.zachleat.com/web/comprehensive-webfonts/) to get a good understanding of the different techniques available. There is also a good [follow-up article](https://css-tricks.com/the-best-font-loading-strategies-and-how-to-execute-them/) that summarises Zach's findings.

### Serve WOFF and WOFF2 font formats

WOFF2 has the superior compression. However, unfortunately [it does not cover all of our Enhanced browsers](https://caniuse.com/woff2) at the time of writing. Therefore consider serving both formats. 

### [font-display CSS](https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-display)

[`font-display`](https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-display) is useful and has the benefit of being a technique that does not require javascript and is easy and straight forward to implement.

`font-display: swap` can be used to achieve FOUT which ensures a user can read text with a fallback font whilst a webfont loads.

> ### Loading fonts with javascript

Reviewer 3: ```suggestion
### Loading fonts with JavaScript
```

Consider using [FontFaceObserver](https://github.com/bramstein/fontfaceobserver) instead of the [CSS Font Loader API](https://drafts.csswg.org/css-font-loading/), as the latter does not have support for all our Enhanced browsers.

### Preloading webfonts

If you have to load a font the conventional way using a `<link>` then consider using the [preload technique](https://web.dev/preload-critical-assets/) to ensure the font is requested as early on in the page load process as possible. 

`<link rel="preload" href="HardingText-Regular-Web.woff2" as="font" type="font/woff2" crossorigin>`

### Self host webfonts if possible

> Some webfonts prevent you from self hosting because of their licensing so check that first. But if you can, consider self hosting a font. As this will potentially reduce the possible points of failure or poor page performance. And it will open up more options to you in terms of possible font loading implementations.

Reviewer 3: ```suggestion
Some webfonts prevent you from self hosting because of their licensing, so check that first. But if you can, consider self hosting a font. This will potentially reduce the possible points of failure or poor page performance, and it will open up more options to you in terms of possible font loading implementations.
```

# Font Loading

## When to use webfonts
Before using webfonts it's worth understanding what benefits you believe a webfont will provide your product and assess those against the potential impact they may have on things like accessibility or page speed. For example:

> - System fonts are battle tested and were produced with legibility in mind. Using webfonts may reduce the readability of the site, especially for people with dyslexia.

Reviewer 4: ```suggestion
- System fonts are extensively tested and produced with legibility in mind. Using webfonts may reduce the readability of the site, especially for people with dyslexia.
```
Want to avoid militaristic language

- Webfonts can negatively impact page performance (additional http requests and increased page weight) especially in markets of strategic importance for Springer Nature like China. This will in turn impact usage numbers.

> Make sure that you feel that on balance the benefit outweighs the impact.

Reviewer 4: ```suggestion
Make sure you can demonstrate that the benefit outweighs the cost.
```
Want to be more prescriptive here

# Font Loading

## When to use webfonts

Webfonts allow us to bring emotion, personality and tone to our sites - which can make them more successful.

However, the task of choosing a webfont and implementing it on your site should be approached cautiously. Webfonts have the potential to impact aspects of your sites such as accessibility and page speed. For example:

> - Using webfonts may reduce the readability of the site, for example for people with dyslexia, or low vision.

Reviewer 3: ```suggestion
- Certain webfonts may reduce the readability of the site, for example for people with dyslexia, or low vision.
```

- Webfonts can negatively impact page performance (additional http requests and increased page weight) especially in markets of strategic importance for Springer Nature like China. This will in turn impact usage numbers.

> Our recommendation is that you use a [system font](https://css-tricks.com/snippets/css/system-font-stack/) instead of a webfont. System fonts are extensively tested, incur no performance penalty, and are produced with legibility in mind.

Reviewer 3: ```suggestion
Our recommendation is that you prefer a [system font](https://css-tricks.com/snippets/css/system-font-stack/) over a webfont. System fonts are extensively tested, incur no performance penalty, and are produced with legibility in mind.
```

Though maybe "prioritise" is a better word than "prefer"?

Author: Yeah I like prioritise better than prefer in that sentence 👍

> Our recommendation is that you use a [system font](https://css-tricks.com/snippets/css/system-font-stack/) instead of a webfont. System fonts are extensively tested, incur no performance penalty, and are produced with legibility in mind.

Reviewer 4: `**We recommend to use a [system font](https://css-tricks.com/snippets/css/system-font-stack/) instead of a webfont**. System fonts are extensively tested, and incur no performance penalty.`

Highlighting the recommendation.

Also taking into account @foxintherain's feedback about Garamond and legibility, we can't say that "system fonts are legible", per se.

Reviewer 5: Garamond isn't a system font though

Reviewer 4: This might be a pertinent point, knew I should've stuck with Webdings 😂

Author: It's probably true though that not all system fonts could be said to have good legibility. So safest to take that bit out I think.

Reviewer 6: My jokes aside about Garamond - I think the main system fonts for OS' (Windows, Apple, Google) absolutely have been tested and very legible - so I think you can leave it in! Maybe I'm getting confused here between browser safe and system os

Author: I dunno, I can't verify that all system fonts meet that sweeping statement of being "produced with legibility in mind" so it does feel safer just to remove. Think it's better without.

> If a webfont is needed then close collaboration between User Experience and Accessibility specialists should be sought to ensure the webfont chosen is satisfactory from both typographic and accessibility perspectives.

Reviewer 3: ```suggestion
If you need a webfont, then close collaboration between User Experience and Accessibility specialists should be sought to ensure your chosen webfont is satisfactory from both typographic and accessibility perspectives.
```

> If a webfont is needed then close collaboration between User Experience and Accessibility specialists should be sought to ensure the webfont chosen is satisfactory from both typographic and accessibility perspectives.

Reviewer 4: `If a webfont is needed then close collaboration between user experience, accessibility, and web performance specialists must be sought to ensure the choice is satisfactory from typographic, accessibility, and performance perspectives.

As a Frontend Developer, it's your role to provide technical guidance and feedback, not to choose the font.`

Changed the "should seek" to a "must seek", included web performance, and added a bit of guidance around the role a FED plays in this.

- [Graded Browser Support](https://github.com/springernature/frontend-playbook/blob/main/practices/graded-browser-support.md)
  - Is the implementation supported by all "enhanced" browsers? If not, is there a good justification for selecting it anyway?
  - Does the implementation work in conjunction with the Cut The Mustard technique?
- [Progressive Enhancement](https://github.com/springernature/frontend-playbook/blob/main/practices/progressive-enhancement.md)
  - This is not meant to force you to choose a non-js implementation, but rather for you to ensure you consider what will happen for users in varying circumstances from the beginning.
- [Our Performance Goals](https://github.com/springernature/frontend-playbook/blob/main/performance/performance-checklist.md)
  - What impact will the implementation have on the page performance of your site? 
  
## Quick Tips
Here are some quick suggestions for techniques that may prove useful:

### Understand the different user experiences in font loading: FOIT, FOUT and FOIT

> - FOIT means _flash of invisible text_. Text on the page is hidden while webfonts are loading, so users will see a page with images but no text. We only show the text when web fonts are loaded.

Reviewer 7: :twitching_eye_emoji: web font -> webfont

> - FOUT means _flash of unstyled text_. When web fonts are loading, we show users a system font. When the web font gets loaded, we change the text back to the desired web font.

Reviewer 7: :twitching_eye_emoji: web font -> webfont

We believe that FOIT should be avoided as it is the worst experience for the user.

### Understand the different implementations available

We highly recommend reading [Zach Leat's comprehensive guide](https://www.zachleat.com/web/comprehensive-webfonts/) to get a good understanding of the different techniques available. There is also a good [follow-up article](https://css-tricks.com/the-best-font-loading-strategies-and-how-to-execute-them/) that summarises Zach's findings.

### Serve WOFF and WOFF2 font formats

WOFF2 has the superior compression. However, unfortunately [it does not cover all of our Enhanced browsers](https://caniuse.com/woff2) at the time of writing. Therefore consider serving both formats. 

### [font-display CSS](https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-display)

[`font-display`](https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-display) is useful and has the benefit of being a technique that does not require JavaScript and is easy and straight forward to implement.

> `font-display: swap` can be used to achieve FOUT which ensures a user can read text with a fallback font whilst a webfont loads.

Reviewer: I know this paragraph is not saying that `font-display: swap` is the best or the only useful value for `font-display`, but given that it's the only one mentioned, I feel like we may want to clarify that all possible `font-display` values have serious downsides. Although we're risking this PR becoming comically large if we include that.

For example, `swap` (and `block` too, I think) will cause the whole page to be rendered completely twice, doubling the amount of paint events and increasing. The geometry is probably not the same between fonts, so the shifts in the layout may be annoying or jarring for some people.

Gov.uk uses `fallback`, and, there's an excellent article from Matt Hobbs showing the whole decision process of why they decided to use that value:
https://calendar.perfplanet.com/2020/a-font-display-setting-for-slow-connections/

I'm not saying that `fallback` is best by the way, just that this is a bit too complex for a short mention so maybe we can just let others do all the hard work and link to that post, or a similar one?

## Implementation

We do not prescribe a particular implementation. Each team should determine which implementation is best suited for them. However, before selecting an implementation a team should be mindful of our working practices such as:

- [Graded Browser Support](https://github.com/springernature/frontend-playbook/blob/main/practices/graded-browser-support.md)
  - Is the implementation supported by all "enhanced" browsers? If not, is there a good justification for selecting it anyway?
  - Does the implementation work in conjunction with the Cut The Mustard technique?
- [Progressive Enhancement](https://github.com/springernature/frontend-playbook/blob/main/practices/progressive-enhancement.md)
  - This is not meant to force you to choose a non-js implementation, but rather for you to ensure you consider what will happen for users in varying circumstances from the beginning.
- [Our Performance Goals](https://github.com/springernature/frontend-playbook/blob/main/performance/performance-checklist.md)
  - What impact will the implementation have on the page performance of your site? 
  
## Quick Tips
Here are some quick suggestions for techniques that may prove useful:

> ### Understand the different user experiences in font loading: FOIT, FOUT and FOIT

Reviewer 8: Typo

> ### Understand the different user experiences in font loading: FOIT, FOUT and FOIT

Reviewer 8: Typo - second **FOIT** should be **FOFT**

Author: Good spot. 😂 
It's amazing how many times you can read something and not spot an obvious mistake like that.
